src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.Values.StringSegment.cs

```diff
@@ -0,0 +1,1109 @@
+
+                // 4-byte sequence containing < 4 bytes
+                [0b11110_100],
+                [0b11110_100, 0b10_001111],
+                [0b11110_100, 0b10_001111, 0b10_111111],
+                
+                // 4-byte overlong
+                [0b11110_000, 0b10_000000, 0b10_000000, 0b10_000000],
+
+                // Greater than Unicode max value
+                [0b11110_111, 0b10_000000],
+                [0b11110_100, 0b10_100000, 0b10_000000],
+            ];
+
+            // Separate each case with a character
+            yield return input.SelectMany(arr => arr.Concat([(byte)'j'])).ToArray();
+
+            // Test without separation
+            yield return input.SelectMany(arr => arr).ToArray();
+        }
+
+        public static IEnumerable<object[]> InvalidUtf8DataWithOptions_TestData =>
+            from data in InvalidUtf8Data()
+            from option in BasicStringJsonOptions
+            select new object[] { data, option };
+
+        [Theory]
+        [MemberData(nameof(InvalidUtf8DataWithOptions_TestData))]
+        public static void WriteStringValueSegment_Utf8_SplitCodePointsReplacement(byte[] inputArr, JsonWriterOptions options)
+        {
+            var expectedBytes = new byte[inputArr.Length * MaxExpansionFactorWhileEscaping];
+
+            options.Encoder.EncodeUtf8(inputArr, expectedBytes, out int bytesConsumed, out int bytesWritten);
+            Assert.Equal(inputArr.Length, bytesConsumed);
+
+            string expectedString = $@"""{Encoding.UTF8.GetString(expectedBytes, 0, bytesWritten)}""";
+
+            SplitCodePointsHelper(inputArr, expectedString, options);
+        }
+
+        private static void SplitCodePointsHelper<T>(
+            T[] inputArr,
+            string expected,
+            JsonWriterOptions options)
+            where T : struct
+        {
+            SplitCodePointsHelper<T>(inputArr, options, output => JsonTestHelper.AssertContents(expected, output));
+        }
+
+        private static void SplitCodePointsHelper<T>(
+            T[] inputArr,
+            JsonWriterOptions options,
+            Action<ArrayBufferWriter<byte>> assert)
+            where T : struct
+        {
+            SplitCodePointsHelper<T>(inputArr.AsSpan(), options, assert);
+        }
+
+        private static void SplitCodePointsHelper<T>(
+            ReadOnlySpan<T> inputArr,
+            JsonWriterOptions options,
+            Action<ArrayBufferWriter<byte>> assert)
+            where T : struct
+        {
+            ReadOnlySpan<T> input = inputArr;
+
+            // Sanity check with non-segmented API
+            {
+                var output = new ArrayBufferWriter<byte>(1024);
+
+                using (var writer = new Utf8JsonWriter(output, options))
+                {
+                    WriteStringValueHelper(writer, input);
+                    writer.Flush();
+                }
+
+                assert(output);
```

Comment:
Thought I was looking at C code for a moment there :-)

Reply:
😆